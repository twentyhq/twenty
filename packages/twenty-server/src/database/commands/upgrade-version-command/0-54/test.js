const scannedPositions = new Set();
let biggestPosition = -1;

// Sort options by position for consistent processing
const sortedOptions = [
  { name: 'a', position: 2 },
  { name: 'b', position: 1 },
  { name: 'c', position: 1 },
  { name: 'd', position: 2 },
].sort((a, b) => a.position - b.position);

for (const option of sortedOptions) {
  if (scannedPositions.has(option.position)) {
    option.position = biggestPosition + 1;
  }

  biggestPosition = Math.max(biggestPosition, option.position);
  scannedPositions.add(option.position);
}

console.log('Sorted options:', sortedOptions);
