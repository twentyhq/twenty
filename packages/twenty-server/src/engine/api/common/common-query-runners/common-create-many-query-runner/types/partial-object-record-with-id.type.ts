import { type ObjectRecord } from 'twenty-shared/types';

export type PartialObjectRecordWithId = Partial<ObjectRecord> & { id: string };
