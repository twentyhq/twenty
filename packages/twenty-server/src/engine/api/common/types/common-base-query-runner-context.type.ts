import { type ObjectLiteral } from 'typeorm';

import { type CommonQueryRunnerOptions } from 'src/engine/api/common/interfaces/common-query-runner-options.interface';
import { type ResolverArgs } from 'src/engine/api/graphql/workspace-resolver-builder/interfaces/workspace-resolvers-builder.interface';

import { type CommonSelectedFieldsResult } from 'src/engine/api/common/types/common-selected-fields-result.type';
import { type WorkspaceDataSource } from 'src/engine/twenty-orm/datasource/workspace.datasource';
import { type WorkspaceRepository } from 'src/engine/twenty-orm/repository/workspace.repository';

export type CommonBaseQueryRunnerContext<Input extends ResolverArgs> = {
  args: Input;
  options: CommonQueryRunnerOptions;
  workspaceDataSource: WorkspaceDataSource;
  repository: WorkspaceRepository<ObjectLiteral>;
  selectedFieldsResult: CommonSelectedFieldsResult;
  isExecutedByApiKey: boolean;
  roleId?: string;
  shouldBypassPermissionChecks: boolean;
};
