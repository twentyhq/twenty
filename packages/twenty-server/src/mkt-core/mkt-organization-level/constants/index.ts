export * from './permission-templates.constants';
export * from './rbac-definitions.constants';
export * from './validation.constants';

export enum OrganizationHierarchyLevel {
  MIN = 1,
  MAX = 8,
}
