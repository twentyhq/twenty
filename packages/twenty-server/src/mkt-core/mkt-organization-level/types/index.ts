export * from './organization-level.types';
export * from './permissions.type';
