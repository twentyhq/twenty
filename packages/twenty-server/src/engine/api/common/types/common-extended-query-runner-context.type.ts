import { type ObjectLiteral } from 'typeorm';

import { type WorkspaceAuthContext } from 'src/engine/api/common/interfaces/workspace-auth-context.interface';

import { type CommonBaseQueryRunnerContext } from 'src/engine/api/common/types/common-base-query-runner-context.type';
import { type GraphqlQueryParser } from 'src/engine/api/graphql/graphql-query-runner/graphql-query-parsers/graphql-query.parser';
import { type WorkspaceDataSource } from 'src/engine/twenty-orm/datasource/workspace.datasource';
import { type WorkspaceRepository } from 'src/engine/twenty-orm/repository/workspace.repository';
import { type RolePermissionConfig } from 'src/engine/twenty-orm/types/role-permission-config';

export type CommonExtendedQueryRunnerContext = Omit<
  CommonBaseQueryRunnerContext,
  'authContext'
> & {
  authContext: WorkspaceAuthContext;
  rolePermissionConfig: RolePermissionConfig;
  repository: WorkspaceRepository<ObjectLiteral>;
  commonQueryParser: GraphqlQueryParser;
  workspaceDataSource: WorkspaceDataSource;
};
