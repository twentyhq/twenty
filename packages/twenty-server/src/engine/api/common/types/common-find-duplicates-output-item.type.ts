import { type ObjectRecord } from 'twenty-shared/types';

export type CommonFindDuplicatesOutputItem = {
  records: ObjectRecord[];
  totalCount: number;
  hasNextPage: boolean;
  hasPreviousPage: boolean;
  startCursor: string | null;
  endCursor: string | null;
};
