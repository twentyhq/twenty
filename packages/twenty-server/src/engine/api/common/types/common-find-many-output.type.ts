import { type ObjectRecord } from 'twenty-shared/types';

import { type CommonPageInfo } from 'src/engine/api/common/types/common-page-info.type';
import { type CommonSelectedFieldsResult } from 'src/engine/api/common/types/common-selected-fields-result.type';

export type CommonFindManyOutput = {
  records: ObjectRecord[];
  aggregatedValues: Record<string, number>;
  totalCount: number;
  pageInfo: CommonPageInfo;
  selectedFieldsResult: CommonSelectedFieldsResult;
};
