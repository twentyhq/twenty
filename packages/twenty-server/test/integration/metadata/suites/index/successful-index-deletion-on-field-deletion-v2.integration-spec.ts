import { deleteOneObjectMetadata } from 'test/integration/metadata/suites/object-metadata/utils/delete-one-object-metadata.util';
import { findManyObjectMetadata } from 'test/integration/metadata/suites/object-metadata/utils/find-many-object-metadata.util';
import { updateOneObjectMetadata } from 'test/integration/metadata/suites/object-metadata/utils/update-one-object-metadata.util';
import { updateFeatureFlag } from 'test/integration/metadata/suites/utils/update-feature-flag.util';
import { jestExpectToBeDefined } from 'test/utils/expect-to-be-defined.util.test';
import { FieldMetadataType } from 'twenty-shared/types';

import { FeatureFlagKey } from 'src/engine/core-modules/feature-flag/enums/feature-flag-key.enum';
import { type FieldMetadataDTO } from 'src/engine/metadata-modules/field-metadata/dtos/field-metadata.dto';
import { type IndexFieldMetadataDTO } from 'src/engine/metadata-modules/index-metadata/dtos/index-field-metadata.dto';
import { type IndexMetadataDTO } from 'src/engine/metadata-modules/index-metadata/dtos/index-metadata.dto';
import { type ObjectMetadataDTO } from 'src/engine/metadata-modules/object-metadata/dtos/object-metadata.dto';
import { deleteOneFieldMetadata } from 'test/integration/metadata/suites/field-metadata/utils/delete-one-field-metadata.util';
import { updateOneFieldMetadata } from 'test/integration/metadata/suites/field-metadata/utils/update-one-field-metadata.util';
import { CUSTOM_OBJECT_DISHES } from 'test/integration/metadata/suites/object-metadata/constants/custom-object-dishes.constants';
import { createOneObjectMetadata } from 'test/integration/metadata/suites/object-metadata/utils/create-one-object-metadata.util';

const findObjectWithIndex = async ({
  objectMetadataId,
}: {
  objectMetadataId: string;
}) => {
  const { objects } = await findManyObjectMetadata({
    expectToFail: false,
    input: {
      filter: {
        id: {
          eq: objectMetadataId,
        },
      },
      paging: {
        first: 1,
      },
    },
    gqlFields: `
        id
        nameSingular
        fieldsList {
          id
          type
        }
        indexMetadataList {
          name
          isUnique
          isCustom
          indexType
          indexFieldMetadataList {
            id
            fieldMetadataId
            createdAt
            updatedAt
            order
          }
        }
      `,
  });

  const foundObject = objects.find(
    (object) => object.id === objectMetadataId,
  ) as ObjectMetadataDTO & {
    fieldsList: FieldMetadataDTO[];
    indexMetadataList: Array<
      IndexMetadataDTO & {
        indexFieldMetadataList: IndexFieldMetadataDTO[];
      }
    >;
  };

  jestExpectToBeDefined(foundObject);
  expect(foundObject.id).toBe(objectMetadataId);
  return foundObject;
};

describe('Index metadata deletion through  creation v2', () => {
  let createdObjectId: string;

  beforeAll(async () => {
    await updateFeatureFlag({
      expectToFail: false,
      featureFlag: FeatureFlagKey.IS_WORKSPACE_MIGRATION_V2_ENABLED,
      value: true,
    });
  });

  afterAll(async () => {
    await updateFeatureFlag({
      expectToFail: false,
      featureFlag: FeatureFlagKey.IS_WORKSPACE_MIGRATION_V2_ENABLED,
      value: false,
    });
  });

  beforeEach(async () => {
    const {
      labelPlural,
      description,
      labelSingular,
      namePlural,
      nameSingular,
    } = CUSTOM_OBJECT_DISHES;

    const { data } = await createOneObjectMetadata({
      expectToFail: false,
      input: {
        labelPlural,
        description,
        labelSingular,
        namePlural,
        nameSingular,
        icon: 'IconTest',
        isLabelSyncedWithName: false,
      },
      gqlFields: `
        id
      `,
    });

    createdObjectId = data.createOneObject.id;
  });

  afterEach(async () => {
    await updateOneObjectMetadata({
      input: {
        idToUpdate: createdObjectId,
        updatePayload: {
          isActive: false,
        },
      },
      expectToFail: false,
    });
    await deleteOneObjectMetadata({
      expectToFail: false,
      input: { idToDelete: createdObjectId },
    });
  });

  it('Should deleted index when all sub fields metadata gets deleted', async () => {
    const foundObject = await findObjectWithIndex({
      objectMetadataId: createdObjectId,
    });
    const tsVectorField = foundObject.fieldsList.find(
      (field) => field.type === FieldMetadataType.TS_VECTOR,
    );

    jestExpectToBeDefined(tsVectorField);

    expect(foundObject.indexMetadataList.length).toBe(1);
    const { indexFieldMetadataList, ...index } =
      foundObject.indexMetadataList[0];

    expect(index).toMatchSnapshot();
    expect(indexFieldMetadataList.length).toBe(1);
    expect(indexFieldMetadataList).toMatchObject([
      {
        id: expect.any(String),
        fieldMetadataId: tsVectorField.id,
        order: 0,
      },
    ]);

    await updateOneFieldMetadata({
      expectToFail: false,
      input: {
        idToUpdate: tsVectorField.id,
        updatePayload: {
          isActive: false,
        },
      },
    });

    await deleteOneFieldMetadata({
      expectToFail: false,
      input: {
        idToDelete: tsVectorField.id,
      },
    });

    {
      const foundObject = await findObjectWithIndex({
        objectMetadataId: createdObjectId,
      });
      const tsVectorField = foundObject.fieldsList.find(
        (field) => field.type === FieldMetadataType.TS_VECTOR,
      );

      expect(tsVectorField).toBeUndefined();

      expect(foundObject.indexMetadataList.length).toBe(0);
    }
  });
});
