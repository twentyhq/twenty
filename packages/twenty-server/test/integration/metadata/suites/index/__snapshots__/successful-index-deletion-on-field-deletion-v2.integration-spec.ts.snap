// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index metadata deletion through  creation v2 Should deleted index when all sub fields metadata gets deleted 1`] = `
{
  "indexType": "GIN",
  "isCustom": false,
  "isUnique": false,
  "name": "IDX_be6fe08944beb9886fb83dfbbf1",
}
`;
